File: moling-server/app/router/template.py
"""墨灵 (Moling) — 项目模板 API 路由。

实现列出模板、获取模板详情、创建/更新/删除模板、使用模板创建项目等端点。
"""

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.dependencies import get_db, get_current_user
from app.service.template_service import template_service
from app.schemas.template import TemplateResp, CreateTemplateReq, UpdateTemplateReq, TemplateListResp, CreateProjectFromTemplateResp
from app.models.user import User

router = APIRouter()


@router.get("", response_model=TemplateListResp)
async def list_templates(
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    genre: str = Query(None, description="按题材筛选"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TemplateListResp:
    """获取模板列表。"""
    result = await template_service.list_templates(
        db,
        page=page,
        page_size=page_size,
        genre=genre,
    )
    return result


@router.get("/{template_id}", response_model=TemplateResp)
async def get_template(
    template_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TemplateResp:
    """获取模板详情。"""
    result = await template_service.get_template(db, template_id)
    return result


@router.post("", response_model=TemplateResp, status_code=201)
async def create_template(
    req: CreateTemplateReq,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TemplateResp:
    """创建新模板（需要登录）。"""
    result = await template_service.create_template(db, req)
    return result


@router.put("/{template_id}", response_model=TemplateResp)
async def update_template(
    template_id: int,
    req: UpdateTemplateReq,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> TemplateResp:
    """更新模板（需要登录）。"""
    result = await template_service.update_template(db, template_id, req)
    return result


@router.delete("/{template_id}", status_code=204)
async def delete_template(
    template_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> None:
    """删除模板（需要登录）。"""
    await template_service.delete_template(db, template_id)


@router.post("/{template_id}/create-project", response_model=CreateProjectFromTemplateResp, status_code=201)
async def create_project_from_template(
    template_id: int,
    title: str = Query(..., description="项目标题"),
    author: str = Query(None, description="作者"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> CreateProjectFromTemplateResp:
    """使用模板创建新项目。"""
    result = await template_service.create_project_from_template(
        db,
        str(current_user.id),
        template_id,
        title=title,
        author=author,
    )
    return result

File: moling-server-rs/crates/moling-api/src/routes/template.rs
//! Template routes — CRUD + create project from template.
//!
//! Mirrors Python `app/router/template.py`.

use axum::{
    extract::{Path, Query, State},
    routing::{get, post},
    Json, Router,
};
use moling_core::error::{AppError, AppResult};
use moling_db::dao::template_dao::TemplateDao;
use moling_auth::CurrentUser;
use sea_orm::Set;
use crate::state::AppState;
use crate::types::*;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/", post(create_template).get(list_templates))
        .route("/{id}", get(get_template).put(update_template).delete(delete_template))
        .route("/{template_id}/create-project", post(create_project_from_template))
}

/// Query params for template listing.
#[derive(Debug, serde::Deserialize)]
struct TemplateListQuery {
    #[serde(default = "default_page")]
    page: u32,
    #[serde(default = "default_page_size")]
    page_size: u32,
    genre: Option<String>,
}

/// POST /templates/ — create a new template.
async fn create_template(
    State(state): State<AppState>,
    _user: CurrentUser,
    Json(req): Json<CreateTemplateReq>,
) -> AppResult<Json<serde_json::Value>> {
    let dao = TemplateDao;
    let id = uuid::Uuid::new_v4().to_string();
    let model = moling_db::entities::template::ActiveModel {
        id: Set(id),
        name: Set(req.name),
        description: Set(req.description),
        genre: Set(req.genre),
        structure: Set(req.structure),
        created_by: Set(Some(_user.user_id.to_string())),
        ..Default::default()
    };
    let t = dao.create(&state.db, model).await?;
    Ok(Json(serde_json::to_value(t)?))
}

/// GET /templates/ — list templates with pagination and optional genre filter.
///
/// Mirrors Python `list_templates`.
async fn list_templates(
    State(state): State<AppState>,
    _user: CurrentUser,
    Query(q): Query<TemplateListQuery>,
) -> AppResult<Json<serde_json::Value>> {
    let skip = (q.page as u64 - 1) * q.page_size as u64;
    let genre = q.genre.as_deref().unwrap_or("");
    let items = TemplateDao.list_by_genre(&state.db, genre, skip, q.page_size as u64).await?;
    let total = TemplateDao.count_by_genre(&state.db, genre).await?;
    Ok(Json(serde_json::json!({
        "items": items,
        "total": total,
        "page": q.page,
        "page_size": q.page_size,
    })))
}

/// GET /templates/{id} — get a template by ID.
async fn get_template(
    State(state): State<AppState>,
    _user: CurrentUser,
    Path(id): Path<String>,
) -> AppResult<Json<serde_json::Value>> {
    let t = TemplateDao.find_by_id(&state.db, &id).await?.ok_or_else(|| AppError::not_found("模板不存在".to_owned()))?;
    Ok(Json(serde_json::to_value(t)?))
}

/// PUT /templates/{id} — update a template.
async fn update_template(
    State(state): State<AppState>,
    _user: CurrentUser,
    Path(id): Path<String>,
    Json(req): Json<UpdateTemplateReq>,
) -> AppResult<Json<serde_json::Value>> {
    use sea_orm::{ActiveModelTrait, IntoActiveModel};
    let t = TemplateDao.find_by_id(&state.db, &id).await?.ok_or_else(|| AppError::not_found("模板不存在".to_owned()))?;
    let mut a = t.into_active_model();
    if let Some(v) = req.name { a.name = Set(v); }
    if let Some(v) = req.description { a.description = Set(v); }
    if let Some(v) = req.genre { a.genre = Set(v); }
    if let Some(v) = req.structure { a.structure = Set(Some(v)); }
    let u = a.update(&state.db).await.map_err(|_| AppError::internal("更新模板失败".to_owned()))?;
    Ok(Json(serde_json::to_value(u)?))
}

/// DELETE /templates/{id} — delete a template.
async fn delete_template(
    State(state): State<AppState>,
    _user: CurrentUser,
    Path(id): Path<String>,
) -> AppResult<axum::http::StatusCode> {
    TemplateDao.delete(&state.db, &id).await?;
    Ok(axum::http::StatusCode::NO_CONTENT)
}

/// POST /templates/{template_id}/create-project — create a project from a template.
///
/// Mirrors Python `create_project_from_template` (template_id in path).
async fn create_project_from_template(
    State(state): State<AppState>,
    _user: CurrentUser,
    Path(template_id): Path<String>,
    Query(req): Query<CreateProjectFromTemplateQuery>,
) -> AppResult<Json<serde_json::Value>> {
    let tmpl = TemplateDao.find_by_id(&state.db, &template_id).await?.ok_or_else(|| AppError::not_found("模板不存在".to_owned()))?;
    let dao = moling_db::dao::project_dao::ProjectDao;
    let model = moling_db::entities::project::ActiveModel {
        user_id: Set(_user.user_id.to_string()),
        title: Set(req.title),
        author: Set(req.author.unwrap_or_default()),
        genre: Set(tmpl.genre),
        creation_mode: Set("from_template".into()),
        template_id: Set(None),
        status: Set("draft".into()),
        word_count: Set(0),
        ..Default::default()
    };
    let p = dao.create(&state.db, model).await?;
    Ok(Json(serde_json::to_value(p)?))
}

/// Query for create-project-from-template.
#[derive(Debug, serde::Deserialize)]
struct CreateProjectFromTemplateQuery {
    title: String,
    author: Option<String>,
}
